### ISSUE TITLE
Implement proper per-period training for walk-forward backtesting to eliminate look-ahead bias

### ISSUE THREAD

**Issue by alexoms** [OptimalMatch/temporal-trading-agents#116]
# Implement Proper Per-Period Training for Walk-Forward Backtesting

## Problem Statement

The current walk-forward backtesting implementation has **look-ahead bias** that invalidates backtest results. When `inference_mode=false`, the system trains ONE model at the beginning using the most recent data, then reuses it for all walk-forward windows - including historical dates that shouldn't have access to future data.

### Current Behavior (Incorrect)

```
Backtest period: 01/01/2023 - 11/15/2025

Step 1: Train model using data from 11/15/2023 - 11/15/2025 (most recent 2y)
Step 2: Apply SAME model to all windows:
  - Window 1: 01/01/2023 - 01/31/2023 ❌ (uses future data!)
  - Window 2: 02/01/2023 - 02/28/2023 ❌ (uses future data!)
  - ...
  - Window N: 10/15/2025 - 11/15/2025 ✓ (only this is valid)
```

This is acknowledged in the code:
```python
# backend/backtesting_engine.py:795
# In training mode, we still train once and reuse (TODO: implement per-period training)
```

### Expected Behavior (Correct)

Each walk-forward window should train a model using ONLY data available before that window:

```
Backtest period: 01/01/2023 - 11/15/2025

Window 1 (01/01/2023 - 01/31/2023):
  - Train on: 07/01/2021 - 01/01/2023 (547 days before window)
  - Test on: 01/01/2023 - 01/31/2023 ✓

Window 2 (02/01/2023 - 02/28/2023):
  - Train on: 08/01/2021 - 02/01/2023 (547 days before window)
  - Test on: 02/01/2023 - 02/28/2023 ✓

Window N (10/15/2025 - 11/15/2025):
  - Train on: 04/15/2024 - 10/15/2025 (547 days before window)
  - Test on: 10/15/2025 - 11/15/2025 ✓
```

### 1. Model Training Architecture

**Location**: `backend/backtesting_engine.py:796-820`

Currently trains once and reuses:
```python
if not self.config.inference_mode:
    # Training mode - train model once at the beginning
    stats, _ = train_ensemble(
        symbol=self.config.symbol,
        ...
        train_window_days=self.config.walk_forward.train_window_days
    )
    self.consensus_stats = stats  # Reused for ALL periods
```

**Should be**: Train a fresh model for each walk-forward window inside the period loop.

### 2. Data Re-Fetching Is Inefficient

**Location**: `examples/crypto_ensemble_forecast.py:213`

Currently fetches data inside training function:
```python
def train_ensemble_model(symbol, period, lookback, ...):
    df = fetch_crypto_data(symbol, period=period, interval=interval)  # ❌ Fetches every time
```

**Issue**: The backtest already has 5 years of price_data loaded. Re-fetching for each window is wasteful.

**Should be**: Accept pre-sliced training data as a parameter, only fetch if not provided.

### Architecture Changes

1. **Per-Window Training Loop**
   - Move model training INSIDE the walk-forward period loop
   - Train a fresh model for each window using data up to `window_start_date`
   - Use ephemeral models (train, use, discard) to minimize memory usage

2. **Data Slicing Instead of Re-Fetching**
   - Backtest already has full historical `price_data` (e.g., 5 years from cache)
   - For each window, slice the existing data: `price_data[window_start - train_window_days : window_start]`
   - Pass sliced data to training function as `training_data` parameter
   - Only fetch if `training_data` not provided (fallback for standalone training)

3. **Prediction Caching for Auto-Optimization** (CRITICAL FOR SPEED)
   - Cache model predictions, not just models
   - Reuse predictions across all parameter combinations
   - **10-50x speedup** for auto-optimization runs

4. **Model Storage Strategy**
   - **Ephemeral approach**: Train, use, discard immediately
   - No need to cache historical window models (only used once)
   - Live/paper trading uses separate model trained on current data

#### Step 1: Update Training Function to Accept Pre-Sliced Data

**File**: `examples/crypto_ensemble_forecast.py`

```python
def train_ensemble_model(
    symbol,
    period,
    lookback,
    forecast_horizon,
    epochs,
    focus,
    model_name,
    interval='1d',
    use_cache=True,
    max_cache_age_hours=6.0,
    fine_tune_epochs=3,
    train_window_days=None,
    training_data=None  # NEW: Accept pre-sliced data
):
    """
    Args:
        training_data: Optional pre-sliced DataFrame for training.
                       If provided, skips data fetching entirely.
                       If None, falls back to fetch_crypto_data().
    """
    from strategies.model_cache import get_model_cache
    from datetime import datetime

    # Auto-calculate period if needed (only used if training_data is None)
    if train_window_days and not training_data:
        fetch_days = int(train_window_days * 1.25)
        if fetch_days <= 180:
            period = '6mo'
        elif fetch_days <= 365:
            period = '1y'
        elif fetch_days <= 730:
            period = '2y'
        else:
            period = '5y'
        print(f"📊 Auto-calculated period: {period} (train_window_days={train_window_days}, fetch_days={fetch_days})")

    # Fetch data (skip if training_data provided)
    if training_data is not None:
        df = training_data  # ✅ Use provided slice
        print(f"✓ Using provided training data: {len(df)} data points")
    else:
        df = fetch_crypto_data(symbol, period=period, interval=interval)  # Fallback
        print(f"✓ Fetched {len(df)} data points ({interval} interval)")

    # ... rest of training logic unchanged
```

**Propagate** `training_data` through:
- `strategies/strategy_utils.py::train_ensemble()`

#### Step 2: Implement Per-Period Training with Data Slicing

**File**: `backend/backtesting_engine.py`

```python
def run_walkforward_backtest(self, price_data: pd.DataFrame, run_id: str) -> BacktestRun:
    """
    Run walk-forward validation backtest.

    Args:
        price_data: Full historical price data (e.g., 5 years from cache)
    """
    validator = WalkForwardValidator(self.config.walk_forward)
    splits = validator.generate_splits(price_data)

    # Load ensemble module once
    ensemble = load_ensemble_module("examples/crypto_ensemble_forecast.py")

    # REMOVE: Training once at the beginning
    # if not self.config.inference_mode:
    #     stats, _ = train_ensemble(...)

    # NEW: Train per-period with data slicing
    for i, (train_df, test_df) in enumerate(splits):
        window_start_date = pd.to_datetime(test_df.iloc[0]['date'])

        if not self.config.inference_mode:
            # Slice training data from existing price_data (no re-fetching!)
            train_cutoff_date = window_start_date - timedelta(days=self.config.walk_forward.train_window_days)

            # Get price data from (cutoff - train_window_days) to cutoff
            # Add buffer for technical indicators
            buffer_days = int(self.config.walk_forward.train_window_days * 0.25)
            train_start_date = train_cutoff_date - timedelta(days=buffer_days)

            # Slice existing price_data (already loaded, no fetching needed)
            training_data_slice = price_data[
                (price_data['date'] >= train_start_date) &
                (price_data['date'] < window_start_date)
            ].copy()

            print(f"\n🔧 Training model for window {i+1}/{len(splits)}")
            print(f"   Window test period: {window_start_date.date()} - {test_df.iloc[-1]['date']}")
            print(f"   Training data: {train_start_date.date()} - {window_start_date.date()} ({len(training_data_slice)} days)")

            # Train ensemble for this window
            configs = get_default_ensemble_configs(horizon=14)
            stats, _ = train_ensemble(
                symbol=self.config.symbol,
                forecast_horizon=14,
                configs=configs,
                name=f"Backtest-Window-{i}",
                ensemble_module=ensemble,
                use_cache=False,  # Don't cache historical window models
                training_data=training_data_slice  # NEW: Pass sliced data
            )
            self.consensus_stats = stats

            # Train multi-horizon models for this window
            multi_horizon_stats = {}
            for horizon in [7, 14, 30]:
                configs_h = get_default_ensemble_configs(horizon=horizon)
                stats_h, _ = train_ensemble(
                    symbol=self.config.symbol,
                    forecast_horizon=horizon,
                    configs=configs_h,
                    name=f"Backtest-Window-{i}-Horizon-{horizon}d",
                    ensemble_module=ensemble,
                    use_cache=False,
                    training_data=training_data_slice  # NEW: Pass sliced data
                )
                multi_horizon_stats[horizon] = (stats_h, None)

            self.multi_horizon_stats = multi_horizon_stats if len(multi_horizon_stats) >= 2 else None

            # Ephemeral approach: Models discarded at end of iteration
            # No need to explicitly delete - garbage collected automatically

        # ... rest of period simulation (generate signals, execute trades)
```

#### Step 3: Implement Prediction Caching for Auto-Optimization

**This is the CRITICAL performance optimization** that provides 10-50x speedup.

**Context**: Auto-optimization tests multiple parameter combinations (different buy/sell thresholds). Currently, it retrains models for EACH combination, but the predictions are identical - only thresholds change.

**File**: `backend/backtesting_engine.py`

Add prediction caching layer:

```python
class PredictionCache:
    """Cache model predictions per window to avoid retraining across parameter combinations."""

    def __init__(self):
        self.cache = {}  # {window_index: {horizon: predictions}}

    def get(self, window_index: int, horizon: int):
        """Get cached predictions for window/horizon."""
        return self.cache.get(window_index, {}).get(horizon)

    def set(self, window_index: int, horizon: int, predictions: Dict):
        """Cache predictions for window/horizon."""
        if window_index not in self.cache:
            self.cache[window_index] = {}
        self.cache[window_index][horizon] = predictions

    def clear(self):
        """Clear all cached predictions."""
        self.cache = {}


def run_walkforward_backtest(self, price_data: pd.DataFrame, run_id: str) -> BacktestRun:
    """
    Run walk-forward validation backtest.

    For auto-optimization, predictions are cached per window and reused
    across all parameter combinations for massive speedup.
    """
    validator = WalkForwardValidator(self.config.walk_forward)
    splits = validator.generate_splits(price_data)
    ensemble = load_ensemble_module("examples/crypto_ensemble_forecast.py")

    # Check if this is part of an auto-optimization run
    is_auto_optimize = hasattr(self, '_prediction_cache')

    if not is_auto_optimize:
        # First run or standalone backtest - create cache
        self._prediction_cache = PredictionCache()

    for i, (train_df, test_df) in enumerate(splits):
        window_start_date = pd.to_datetime(test_df.iloc[0]['date'])

        # Check if predictions already cached (from previous parameter combo)
        cached_predictions = self._prediction_cache.get(i, 14)  # horizon=14

        if cached_predictions is not None:
            print(f"✓ Using cached predictions for window {i+1}/{len(splits)} (skipping training)")
            self.consensus_stats = cached_predictions

            # Also load cached multi-horizon predictions
            multi_horizon_stats = {}
            for horizon in [7, 14, 30]:
                cached_h = self._prediction_cache.get(i, horizon)
                if cached_h:
                    multi_horizon_stats[horizon] = (cached_h, None)
            self.multi_horizon_stats = multi_horizon_stats if len(multi_horizon_stats) >= 2 else None

        else:
            # First time seeing this window - train and cache predictions
            print(f"\n🔧 Training model for window {i+1}/{len(splits)} (first run, will cache)")

            # ... (training code from Step 2 above)

            # Cache the predictions for reuse
            self._prediction_cache.set(i, 14, stats)
            for horizon, (stats_h, _) in multi_horizon_stats.items():
                self._prediction_cache.set(i, horizon, stats_h)

        # ... rest of period simulation (generate signals using cached/fresh predictions)
```

**File**: `backend/parameter_optimizer.py`

Update optimizer to share prediction cache across parameter combinations:

```python
def optimize(self, base_config: BacktestConfig, parameter_grid: List[Dict], price_data: pd.DataFrame, metric: str):
    """
    Run optimization across parameter combinations.

    Shares prediction cache across all combinations for 10-50x speedup.
    """
    from backend.backtesting_engine import PredictionCache

    # Create shared prediction cache for all parameter combinations
    shared_cache = PredictionCache()

    results = []
    for i, params in enumerate(parameter_grid):
        print(f"\n{'='*60}")
        print(f"Testing parameter combination {i+1}/{len(parameter_grid)}")
        print(f"Params: {params}")
        print(f"{'='*60}")

        # Update config with parameters
        config = self._apply_params(base_config, params)

        # Run backtest with shared cache
        engine = BacktestEngine(config)
        engine._prediction_cache = shared_cache  # Share cache across all runs

        run_id = str(uuid.uuid4())

        if config.walk_forward.enabled:
            result = engine.run_walkforward_backtest(price_data, run_id)
        else:
            result = engine.run_backtest(price_data, run_id)

        results.append((params, result.metrics, result.run_id))

    # Clear cache after optimization completes
    shared_cache.clear()

    return results
```

**Performance Impact**:
```
Without prediction caching:
  10 param combos × 100 windows × 5 min training = 83 hours

With prediction caching:
  (100 windows × 5 min training) + (10 param combos × 10 sec threshold application) = 8.3 hours

Speedup: 10x (scales with number of parameter combinations)
```

#### Step 4: Model Storage Strategy

**Approach**: Ephemeral models (train, use, discard immediately)

**Rationale**:
- Historical window models are only used once per window
- Not needed for live/paper trading (separate model trained on current data)
- Caching 100+ models wastes disk space (50GB+)
- Prediction caching (Step 3) eliminates need for model retraining

**Implementation**:
- Models are automatically garbage collected at end of each loop iteration
- No explicit caching or storage needed
- Memory usage: ~3GB peak per window (model + data), returns to baseline after discard
- Works on machines with 8GB+ RAM

**For Live Trading**:
- Separate workflow trains ONE model on data up to `now()`
- Cached normally with existing model cache system
- No relationship to historical backtest models

## Testing Plan

1. **Unit Tests**
   - Test `training_data` parameter acceptance in `train_ensemble_model()`
   - Verify data slicing produces correct date ranges
   - Test prediction cache get/set operations

2. **Integration Test - Basic Walk-Forward**
   - Run walk-forward backtest on NVDA (2023-2025) with 10 windows
   - Verify each window trains on correct historical data slice
   - Confirm no future data leakage
   - Check memory usage stays under 5GB

3. **Integration Test - Auto-Optimization**
   - Run auto-optimize with 10 parameter combinations
   - Verify first run trains all windows (e.g., 100 windows × 5 min = 500 min)
   - Verify subsequent runs reuse cached predictions (e.g., <1 min per combo)
   - Measure actual speedup ratio

4. **Validation**
   - Compare results with/without proper per-period training
   - Should see more realistic (likely lower) performance metrics
   - Verify training logs show correct data ranges per window
   - Confirm predictions are identical when using cache

### Single Backtest
- **Training time**: Increases from 1× to N× (where N = number of windows)
- **Example**: 100 windows × 5 min = 500 min (8.3 hours) vs previous 5 min
- **Mitigation**: Acceptable because results are now correct (no look-ahead bias)

### Auto-Optimization (Critical Use Case)
- **Without prediction caching**: 10 combos × 100 windows × 5 min = **83 hours** ❌
- **With prediction caching**: (100 windows × 5 min) + (10 combos × 10 sec) = **8.3 hours** ✅
- **Speedup**: **10x** (scales linearly with number of parameter combinations)

### Memory Usage
- **Peak per window**: ~3GB (model + training data)
- **After discard**: Returns to baseline
- **Total**: No accumulation, same memory footprint throughout run

### Results Accuracy
- **Expected**: More conservative/realistic backtest metrics
- **Reason**: Removes look-ahead bias, models trained only on past data
- **Benefit**: Results better predict live trading performance

## Related Issues

- Originally discussed in context of "insufficient data" errors (#114, #115)
- User expected proper walk-forward validation when setting `train_window_days=547`
- Current implementation trains on future data, invalidating backtest results
- Auto-optimization taking hours due to redundant retraining

## Priority

**High** - This is a fundamental correctness issue that:
1. Invalidates all walk-forward backtest results (look-ahead bias)
2. Makes auto-optimization impractically slow (83 hours without prediction caching)
3. Prevents users from trusting backtest performance metrics

The prediction caching component is equally critical for usability - without it, auto-optimization becomes impractical.

## Files to Modify

1. `examples/crypto_ensemble_forecast.py` - Add `training_data` parameter
2. `strategies/strategy_utils.py` - Propagate `training_data` parameter
3. `backend/backtesting_engine.py` - Implement per-period training + prediction caching
4. `backend/parameter_optimizer.py` - Share prediction cache across parameter combos
5. `backend/models.py` - Add any needed config fields

### Phase 1: Basic Per-Period Training (MVP)
- Accept `training_data` parameter in training functions
- Implement per-window training loop with data slicing
- Use ephemeral models (no caching)
- **Outcome**: Correct walk-forward validation (no look-ahead bias)

### Phase 2: Prediction Caching (Performance)
- Implement `PredictionCache` class
- Cache predictions per window/horizon
- Share cache across parameter combinations in auto-optimizer
- **Outcome**: 10-50x speedup for auto-optimization

### Phase 3: Optimization & Polish
- Add progress indicators for long-running backtests
- Implement memory monitoring/warnings
- Add cache hit/miss metrics to logs
- **Outcome**: Better UX for long-running optimizations
